spyder/app/mainwindow.py

```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+            raise Exception('Plugin "{}" not found!'.format(plugin_name))
+
+    def show_status_message(self, message, timeout):
+        """
+        Show a status message in Spyder Main Window.
+        """
+        status_bar = self.statusBar()
+        if status_bar.isVisible():
+            status_bar.showMessage(message, timeout)
+
+    def show_plugin_compatibility_message(self, message):
+        """
+        Show a compatibility message.
+        """
+        messageBox = QMessageBox(self)
+        messageBox.setWindowModality(Qt.NonModal)
+        messageBox.setAttribute(Qt.WA_DeleteOnClose)
+        messageBox.setWindowTitle(_('Compatibility Check'))
+        messageBox.setText(message)
+        messageBox.setStandardButtons(QMessageBox.Ok)
+        messageBox.show()
+
+    def add_plugin(self, plugin):
+        """
+        Add plugin to plugins dictionary.
+        """
+        self._PLUGINS[plugin.CONF_SECTION] = plugin
+
+    def register_plugin(self, plugin):
+        """
+        Register a plugin in Spyder Main Window.
+        """
+        from spyder.api.plugins import SpyderDockablePlugin
+        self.set_splash(_("Loading {}...".format(plugin.get_name())))
+        logger.info("Loading {}...".format(plugin.NAME))
+
+        # Check plugin compatibility
+        is_compatible, message = plugin.check_compatibility()
+        plugin.is_compatible = is_compatible
+        plugin.get_description()
+
+        if not is_compatible:
+            self.show_compatibility_message(message)
+            return None
+
+        # Signals
+        plugin.sig_free_memory_requested.connect(self.free_memory)
+        plugin.sig_quit_requested.connect(self.close)
+        plugin.sig_redirect_stdio_requested.connect(
+            self.redirect_internalshell_stdio)
+        plugin.sig_status_message_requested.connect(self.show_status_message)
+
+        if isinstance(plugin, SpyderDockablePlugin):
+            plugin.sig_switch_to_plugin_requested.connect(
+                self.switch_to_plugin)
+
+        # Register plugin
+        plugin._register()
+        plugin.register()
+
+        if isinstance(plugin, SpyderDockablePlugin):
+            # Add dockwidget
+            logger.info("Adding dockwidget for {}...".format(plugin.NAME))
+            self.add_dockwidget(plugin)
+
+            # Update margins
+            margin = 0
+            if CONF.get('main', 'use_custom_margin'):
+                margin = CONF.get('main', 'custom_margin')
+            plugin.update_margins(margin)
+
+        # First time plugin starts
+        if plugin.get_conf_option('first_time', True):
+            try:
+                logger.info("Tabify {} dockwidget...".format(plugin.NAME))
+                if isinstance(plugin, SpyderDockablePlugin) and plugin.TABIFY:
+                    next_to_plugin = self.get_plugin(plugin.TABIFY)
+                    self.tabify_plugins(plugin, next_to_plugin)
+            except NotImplementedError:
+                pass

```

Comment:
This is a little more complex at this point and should not hold this PR back. Since the new Spyder API will allow for users to start spyder with only the plugins they want, using a "Plugin" on Tabify will not cut it. Some options:

1.) We make that Tabify Plugin a REQUIRES dependency which feels very odd.
2.) We tabify on "abstract locations" that are defined for the standard layouts
3.) We do not make TABIFY a required plugin, and if not found, we use a default location that will always exist, for example, the Internal Console.

Reply:
For now using option 3